File: Data Structures & Algorithms/max-water-container/submission-0.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        l = 0
        r = len(heights) - 1
        max = 0
        
        while (l < r):
            area = min(heights[l], heights[r]) * (r-l)
            if (area > max):
                max = area
            if (heights[l] >= heights[r]):
                r = r-1
            else:
                l = l+1
        return max

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        // 0: height, 1: index
        Stack<int[]> stack = new Stack<>();

        int maxArea = 0;

        for (int i = 0; i < heights.length; i++) {
            int start = i;
            int height = heights[i];
            while (!stack.isEmpty() && height < stack.peek()[0]) {
                int[] pair = stack.pop();
                maxArea = Math.max(maxArea, pair[0] * (i - pair[1]));
                start = pair[1];
            }
            stack.push(new int[] {height, start});
        }

        int n = heights.length;
        for (int[] pair: stack) {
            maxArea = Math.max(maxArea, pair[0] * (n - pair[1]));
        }

        return maxArea;
    }
}
